### PR TITLE
Detach stream initialisation from sockets

### PR DESCRIPTION
**Pull request by kasperisager**
This PR adds a preconnect hook for when a socket receives a message from an unknown remote stream:

https://github.com/mafintosh/libudx/blob/8f36c19873c667ec4f9323ba4520838a7eb764d3/test/stream.js#L197-L219

### CODE REVIEW

**Comment by kasperisager** [comment]
@mafintosh The tests are awaiting #32.

**Comment by kasperisager** [comment]
@mafintosh This is ready for review.

File: src/udx.c [hunk 1]
```diff
@@ -655,7 +655,10 @@ on_uv_poll (uv_poll_t *handle, int status, int events) {
     return;
   }
 
-  update_poll(socket);
+  // update the poll is the socket is still active.
+  if (uv_is_active((uv_handle_t *) &socket->io_poll)) {
+    update_poll(socket);
+  }
```
Comment (kasperisager) [inline src/udx.c:661]: This was also causing the `destroy streams and close socket in callback` test case to fail on `master`.

File: lib/socket.js [hunk 2]
```diff
@@ -161,6 +161,10 @@ module.exports = class UDX extends events.EventEmitter {
     return this._reqs.push({ handle, buffer: null, onflush: null }) - 1
   }
 
+  static createSocket () {
+    return new UDX()
+  }
+
```
Comment (kasperisager) [inline lib/socket.js:167]: For compatibility with `dgram`.

Comment (mafintosh) [reply lib/socket.js:167]: 👍

File: src/udx.c [hunk 3]
```diff
@@ -417,8 +430,23 @@ process_packet (udx_t *socket, char *buf, ssize_t buf_len) {
   buf += UDX_HEADER_SIZE;
   buf_len -= UDX_HEADER_SIZE;
 
-  udx_stream_t *stream = (udx_stream_t *) udx__cirbuf_get(&(socket->streams_by_id), local_id);
-  if (stream == NULL || stream->status & UDX_STREAM_DEAD) return 0;
+  udx_stream_t *stream = lookup_stream(socket, local_id);
+
+  if (stream == NULL || (stream->status & UDX_STREAM_CONNECTED) == 0) {
+    if (socket->on_preconnect != NULL) {
```
Comment (mafintosh) [inline src/udx.c:436]: Can we add a comment here saying the preconnect hook is to give users a chance to connect the stream if it's not connected, hence we do the logic twice here

File: src/udx.c [hunk 4]
```diff
@@ -417,8 +430,23 @@ process_packet (udx_t *socket, char *buf, ssize_t buf_len) {
...
+      socket->on_preconnect(socket, local_id, addr);
+
+      stream = lookup_stream(socket, local_id);
+
+      if (stream == NULL || (stream->status & UDX_STREAM_CONNECTED) == 0) {
+        return 0;
+      }
+    } else {
+      return 0;
+    }
```
Comment (mafintosh) [inline src/udx.c:446]: negative the if and move this return up, ie

```
if (!cond) return 0;
```

vs

```
if (cond) {
} else {
  return 0;
}

File: src/udx.c [hunk 5]
```diff
@@ -870,28 +887,49 @@ udx_stream_init (udx_t *socket, udx_stream_t *handle, uint32_t *local_id, udx_st
...
 }
 
-// TODO: finish this
-void
+int
 udx_stream_recv_stop (udx_stream_t *handle) {
+  if ((handle->status & UDX_STREAM_RECEIVING) == 0) return 0;
+
   handle->on_recv = NULL;
+  handle->status ^= UDX_STREAM_RECEIVING;
```
Comment (mafintosh) [inline src/udx.c:906]: fancy 🤣

File: src/udx.c [hunk 6]
```diff
@@ -934,20 +972,34 @@ udx_stream_check_timeouts (udx_stream_t *handle) {
...
-void
-udx_stream_connect (udx_stream_t *handle, uint32_t remote_id, const struct sockaddr *remote_addr) {
-  int already_connected = handle->status & UDX_STREAM_CONNECTED;
+int
+udx_stream_connect (udx_stream_t *handle, udx_t *socket, uint32_t remote_id, const struct sockaddr *remote_addr, udx_stream_close_cb close_cb) {
+  if (handle->status & UDX_STREAM_CONNECTED) {
+    return UV_EISCONN;
+  }
+
+  if (socket->streams_len >= 65536) return UV_EBUSY;
```
Comment (mafintosh) [inline src/udx.c:981]: I personally find `if (streams_len > 0xffff)` clearer but that might just be me

**Comment by mafintosh** [comment]
Looks great, only some minor comments

**Comment by mafintosh** [comment]
LGTM

**Pull request merged**